sklearn/pipeline.py

```diff
@@ -75,10 +79,10 @@ class Pipeline(_BaseComposition):
 
     Parameters
     ----------
-    steps : list of tuple
-        List of (name, transform) tuples (implementing `fit`/`transform`) that
-        are chained in sequential order. The last transform must be an
-        estimator.
+    steps : list of tuples
+        List of (name, transform) tuples (implementing `fit` and `transform`)
+        that are chained in sequential order. The last step can only implement
+        `fit`.
```

Comment:
There isn't many requirements on the last step. For example, one can have a stateless transformers in a pipeline:

```python
from sklearn.pipeline import make_pipeline
import numpy as np

class DoubleIt:
    def transform(self, X, y=None):
        return 2*X

X = np.array([[1, 2, 3], [4, 5, 6]])
p = make_pipeline(DoubleIt(), DoubleIt())

p.transform(X)
# array([[ 4,  8, 12],
#       [16, 20, 24]])
```

As long as the methods you call on the pipeline exist in the final step, then everything still works.

Reply:
Thanks @thomasjpfan, I didn't think of that... I will change the wording to also allow for only stateless transformers. :)

But look at this:

```
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
import numpy as np

class DoubleIt:

    def transform(self, X, y=None):
        return 2*X

X = np.array([[1, 2, 3], [4, 5, 6]])
p = Pipeline([
            ('double1', DoubleIt()),
            ('double2', DoubleIt()),
            ('linreg', LinearRegression())     # same results with ('linreg', None), but not with "passthrough"
            ])

print(p.fit(X))
```

Error:
`TypeError: All intermediate steps should be transformers and implement fit and transform or be the string 'passthrough' '<__main__.DoubleIt object at 0x7fbed9d1f310>' (type <class '__main__.DoubleIt'>) doesn't`

Whenever the one of the steps implements a `fit`, the previous steps are expected to be stateful transformers, too. This is because `def _validate_steps() ` only checks the steps if `.fit()` is called.

Should `_validate_steps`()  be modified to be less strict about stateless transformers? And should it then also run on `transform()`, and all the other possible methods?

Reply:
Given where the discussion in https://github.com/scikit-learn/scikit-learn/issues/27014 is headed, I think we need to enforce that `pipeline.fit` works. According to our [estimator docs](https://scikit-learn.org/dev/developers/develop.html#estimators), an estimator must define `fit`. For `pipeline.fit` to work:

- All steps must define `fit`.
- All non-last steps must also define `transform`.